overlord/snapstate/refresh.go

```diff
@@ -0,0 +1,213 @@
+	// sufficient to perform the check, even though individual processes
+	// may fork or exit, we will have per-security-tag information about
+	// what is running.
+	lock, err := snaplock.OpenLock(info.SnapName())
+	if err != nil {
+		return err
+	}
+	defer lock.Close()
+	if err := lock.Lock(); err != nil {
+		return err
+	}
+
+	var busyAppNames []string
+	var busyHookNames []string
+	var busyPIDs []int
+
+	canHookRunDuringRefresh := func(hook *snap.HookInfo) bool {
+		return false
+	}
+
+	for name, app := range info.Apps {
+		if canAppRunDuringRefresh(app) {
+			continue
+		}
+		PIDs, err := pidsOfSecurityTag(app.SecurityTag())
+		if err != nil {
+			return err
+		}
+		if len(PIDs) > 0 {
+			busyAppNames = append(busyAppNames, name)
+			busyPIDs = append(busyPIDs, PIDs...)
+		}
+	}
+
+	for name, hook := range info.Hooks {
+		if canHookRunDuringRefresh(hook) {
+			continue
+		}
+		PIDs, err := pidsOfSecurityTag(hook.SecurityTag())
+		if err != nil {
+			return err
+		}
+		if len(PIDs) > 0 {
+			busyHookNames = append(busyHookNames, name)
+			busyPIDs = append(busyPIDs, PIDs...)
+		}
+	}
+	if len(busyAppNames) == 0 && len(busyHookNames) == 0 {
+		return nil
+	}
+	sort.Strings(busyAppNames)
+	sort.Strings(busyHookNames)
+	sort.Ints(busyPIDs)
+	return &BusySnapError{
+		snapName:      info.SnapName(),
+		busyAppNames:  busyAppNames,
+		busyHookNames: busyHookNames,
+		pids:          busyPIDs,
+	}
+}
+
+// SoftNothingRunningRefreshCheck looks if there are at most only service processes alive.
+//
+// The check is designed to run early in the refresh pipeline. Before
+// downloading or stopping services for the update, we can check that only
+// services are running, that is, that no non-service apps or hooks are
+// currently running.
+//
+// Since services are stopped during the update this provides a good early
+// precondition check. The check is also deliberately racy as existing snap
+// commands can fork new processes or existing processes can die. After the
+// soft check passes the user is free to start snap applications and block the
+// hard check.
+func SoftNothingRunningRefreshCheck(info *snap.Info) error {
+	return genericRefreshCheck(info, func(app *snap.AppInfo) bool {
+		return app.IsService()
+	})
+}
+
+// HardNothingRunningRefreshCheck looks if there are any undesired processes alive.
+//
+// The check is designed to run late in the refresh pipeline, after stopping
+// snap services. At this point non-enduring services should be stopped, hooks
+// should no longer run, and applications should be barred from running
+// externally (e.g. by using a new inhibition mechanism for snap run).
+//
+// The check fails if any process belonging to the snap, apart from services
+// that are enduring refresh, is still alive. If a snap is busy it cannot be
+// refreshed and the refresh process is aborted.
+func HardNothingRunningRefreshCheck(info *snap.Info) error {
+	return genericRefreshCheck(info, func(app *snap.AppInfo) bool {
```

Comment:
The difference between Hard/Soft check is just "RefreshMode == "ensure", yes? This makes me wonder if instead of this indirection and two different exported functions and one generic helper we could simply have a single function and a flag, something like (names are just strawmans):
```
type RefreshCheckFlags {
    HardNothingRunning bool // HardNothingRunning looks if there are any undesired processes alive.
}
func RefreshCheck(info *snap.Info, flags RefreshCheckFlags) error {
...
```
or do we expect that canAppRunDuringRefresh grow soon? If not, maybe YAGNI? But I understand this is partly personal preference so happy to be ignored if you and @pedronis  think this does not really help making things easier to follow.

Reply:
I don't know how this will behave in the future. I would like to keep the two public functions explicitly because it matches our discussions from Cape Town and earlier. The soft and hard checks will still exist even though their details may vary. I don't think this warrants a merging (given that we have the private generic function already).

Reply:
Yes, agreed, I think  the two functions map the way we discussed about this.